Improve Metric Assignment Results

jessie-ross opened the pull request:
<!-- Describe your changes in detail. -->
**This PR improves metric assignment results:**

(Edit: Scrubbed Photos)

With only one data point:

(Edit: Scrubbed Photos)

Changes:
- Adds sections to Metric Assignment Results
- Moves the graphs to the top
- Add variation estimates
- Add interpretation of estimates
- Changes diff graph for conversions to have "percentage points" as unit.
- Refactors metric formatting into `formatMetricValue`
- Additionally changes the Participant graph to use: "lines+markers" to partially fix issue #355 
   (Probably shouldn't be in this PR but leaving it in as it is so small.)

<!-- Remember to include context: Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Partially resolves issue #419 

## How has this been tested?

- Automated tests that cover added/changed functionality
- Manual testing with production data (locally)

jessie-ross commented:
This does reveal the x100 percentage points, so it does need to be done in time with other percentage point changes. It can be taken out pretty easily though.

yanirs commented on src/components/experiments/single-view/overview/MetricAssignmentsPanel.tsx:
```diff
@@ -213,15 +213,10 @@ function MetricAssignmentsPanel({
+                  {formatMetricValue(
+                    resolvedMetricAssignment.minDifference,
+                    resolvedMetricAssignment.metric.parameterType,
+                    true,
```
It's hard to tell what's `true` referring to in this context. It'd be easier to read if `formatMetricValue()` was changed to work with an object as a single parameter.

yanirs commented on src/utils/formatters.ts:
```diff
@@ -1,17 +1,59 @@
  */
-function formatUsCurrencyDollar(value: number): string {
-  return usCurrencyDollarFormatter.format(value)
+export const metricValueFormatData: Record<
```
Why are these two consts exported?

jessie-ross replied:
Ah yeah I was going to use the transform for the graphs but I can unexport it.

yanirs commented:
This feels like a step in the right direction, but I'm not sure about some of these changes:

* Was the change to not always show dollar values with cents intentional (`$x.cc`)? If so, what's the reasoning?
* Always spelling out `pp` feels too verbose. I think that the tooltip approach is better in places where we use % or $. Experimenters may need to hover once or twice, but then they won't need to do it anymore.
* It'd be good to reorder things under the `Latest Estimates` heading, putting the difference first, as it's the most important estimate. It'll also work nicely when we add support for more than two variations. 
* Shall we hide the metric-level graphs in the single-point case? Or are you planning on doing it separately?
* Having the warnings under `Latest Estimates` is a bit unexpected. Maybe they should be under `Analysis Fine Print`?
* I find that my eyes jump a lot with the way the three tables are (mis)aligned. Maybe we can use a single table with subheadings for better alignment of the second column? Something like:
<table>
<tr><td><h3>Latest Estimates</h3></td><td></td></tr>
<tr><td>Difference</td><td>[...]</td></tr>
<tr><td>control</td><td>[...]</td></tr>
<tr><td><h3>Metric Assignment Details</h3></td><td></td></tr>
<tr><td>Metric Description</td><td>This is a metric...</td></tr>
</table>

By the way, I edited the PR description to add the word "issue" before the issue number so that they don't get closed by this PR (`partially fixes issue #xxx` doesn't close the issue, while `partially fixes #xxx` does). 

> This does reveal the x100 percentage points, so it does need to be done in time with other percentage point changes. It can be taken out pretty easily though.

I don't think it's an issue since it's already a problem in the graph and the recommendations. We should aim to fix it this week, though.

jessie-ross commented:
> Was the change to not always show dollar values with cents intentional ($x.cc)? If so, what's the reasoning?

Yeah, since this is ARPU the level of precision you might need could be below 2 decimal places and then I figured we should just use the same level of rounding across Conversion and Revenue (4 decimal places).

> Always spelling out pp feels too verbose. I think that the tooltip approach is better in places where we use % or $. Experimenters may need to hover once or twice, but then they won't need to do it anymore.

Hmm 🤔 

> It'd be good to reorder things under the Latest Estimates heading, putting the difference first, as it's the most important estimate. It'll also work nicely when we add support for more than two variations.

:+1: 

> Shall we hide the metric-level graphs in the single-point case? Or are you planning on doing it separately?

Yeah separately, I accidentally let it creep in here for the participants.

> Having the warnings under Latest Estimates is a bit unexpected. Maybe they should be under Analysis Fine Print?

I think it is important to have warnings by the estimates, I don't think Experimenters are likely to look into the Fine Print.

> I find that my eyes jump a lot with the way the three tables are (mis)aligned. Maybe we can use a single table with subheadings for better alignment of the second column? Something like:

I can just fix the alignment

yanirs commented:
> Yeah, since this is ARPU the level of precision you might need could be below 2 decimal places and then I figured we should just use the same level of rounding across Conversion and Revenue (4 decimal places).

Hmm... Maybe we should just change the default to two decimal places? For percentages, four decimal places is actually six (1% = 0.01), and for dollar values, I believe people don't usually care about fractions of cents.

> I think it is important to have warnings by the estimates, I don't think Experimenters are likely to look into the Fine Print.

That's a good point. Maybe add a visual indicator to differentiate them? Something simple like ":warning: Warnings" may work well.

> I can just fix the alignment

That would work too :slightly_smiling_face:

jessie-ross commented:
## Update

- Changed rounding to 2 digits past the point
- Changed to "pp" with tooltip everywhere - replaced `formatMetricValue` with component `MetricValue`
- Put difference up top
- I've ended up hiding the plots for 1 data point:
<img width="1283" alt="Screen Shot 2021-01-14 at 4 10 00 pm" src="https://user-images.githubusercontent.com/971886/104571531-27eb2a00-568e-11eb-9233-42ee953065e0.png">

- Put the warning emoji in front of the warning line
- Align the table columns

yanirs commented on src/components/general/MetricValue.tsx:
```diff
@@ -0,0 +1,84 @@
+  const format = metricValueFormatData[`${metricParameterType}${isDifference ? '_difference' : ''}`]
+  return (
+    <>
+      {format.prefix}
```
Looking at https://abacus-git-add-improve-metric-assignments-results.a8c.vercel.app/experiments/2-experiment-name-123/results, negative dollar values look odd (`$-0.25` rather than `-$0.25`). If you don't want to use `Intl.NumberFormat` to work around it, maybe a good solution for now would be to make the prefix `USD ` (with a trailing space, perhaps nbsp)? This would also make it less ambiguous, and something like `USD -0.25` would probably be clear enough in a sentence.

yanirs commented on src/components/general/MetricValue.test.tsx:
```diff
@@ -0,0 +1,91 @@
+      
+    </div>
+  `)
+})
```
It'd be good to add two more test cases:
* Negative dollar value (see comment below)
* Many decimal points, e.g., 0.1232455 (validate the rounding)

yanirs commented on src/components/experiments/single-view/results/CondensedLatestAnalyses.tsx:
```diff
@@ -373,31 +386,208 @@ function AnalysisDetailPanel({
+        </div>
+      ) : (
+        <Typography variant='body1' className={classes.noPlotMessage}>
+          (Not enough data to plot.)
```
I think we can just render nothing, or maybe change the wording to: `Past values will be plotted once we have more than one day of results.` (no need for parentheses).

yanirs commented on src/components/experiments/single-view/results/CondensedLatestAnalyses.tsx:
```diff
@@ -373,31 +386,208 @@ function AnalysisDetailPanel({
+          <Plot
+            layout={{
+              ...Visualizations.plotlyLayoutDefault,
+              title: isConversion ? `Conversion rate estimates by variation (%)` : `Revenue estimates by variation ($)`,
```
If we change the metric value prefix to USD (see below), we should also change it here and on line 413.

yanirs commented on src/components/experiments/single-view/results/CondensedLatestAnalyses.tsx:
```diff
@@ -373,31 +386,208 @@ function AnalysisDetailPanel({
+                  <MetricValue
+                    value={latestEstimates[`variation_${variation.variationId}`].top}
+                    metricParameterType={metric.parameterType}
+                  />
```
This should also have a full stop (like the difference sentence) :slightly_smiling_face:

yanirs reviewed (commented):
Looks great! See a few more inline comments.

yanirs reviewed (approved):
:ship:
